File: src/app/(app)/staff/scan-queue.ts
/**
 * Offline scan queue (§6.4).
 *
 * Counter Wi-Fi drops. When it does, the scan is buffered locally and replayed
 * on reconnect rather than lost — the student has already walked away with
 * their food, so the attendance row must eventually exist.
 *
 * Replay is safe because the server is idempotent: the
 * `UNIQUE (tenant_id, student_id, service_date, meal_slot)` constraint means a
 * replayed scan produces `ALREADY_SERVED` and zero additional rows. That is a
 * *success* from the queue's point of view — the record exists, which is all it
 * was trying to achieve.
 *
 * `localStorage` rather than IndexedDB deliberately: the payloads are tiny, the
 * queue is short-lived, and a synchronous API cannot lose a write when the tab
 * is closed mid-transaction.
 */

const STORAGE_KEY = "messos.scanQueue.v1";

/**
 * Beyond this the meal is over and replaying would write the wrong service
 * date, so a stale entry must not be synced.
 *
 * It must not be *discarded* either. These are students who already walked away
 * with their food; a tablet left offline overnight would otherwise erase a
 * whole service with nobody ever knowing. Stale entries are held back from sync
 * and surfaced to staff instead — see `expiredScans`.
 */
export const MAX_QUEUE_AGE_MS = 6 * 60 * 60 * 1000;
const MAX_ENTRIES = 500;

export interface QueuedScan {
  /** Client-generated, so a retry of the same buffered scan is recognisable. */
  readonly id: string;
  readonly queuedAt: number;
  readonly body: Record<string, unknown>;
}

function read(): QueuedScan[] {
  if (typeof window === "undefined") return [];
  try {
    const raw = window.localStorage.getItem(STORAGE_KEY);
    if (!raw) return [];
    const parsed = JSON.parse(raw) as unknown;
    return Array.isArray(parsed) ? (parsed as QueuedScan[]) : [];
  } catch {
    // Corrupt storage must not brick the scanner mid-service.
    return [];
  }
}

function write(entries: QueuedScan[]): void {
  if (typeof window === "undefined") return;
  try {
    window.localStorage.setItem(STORAGE_KEY, JSON.stringify(entries.slice(-MAX_ENTRIES)));
  } catch {
    // Quota exceeded or private mode. Losing the buffer is bad, but throwing
    // here would take the whole scanner down mid-queue.
  }
}

/** Entries still young enough to sync. */
export function queuedScans(): QueuedScan[] {
  const now = Date.now();
  return read().filter((entry) => now - entry.queuedAt < MAX_QUEUE_AGE_MS);
}

/**
 * Entries too old to sync safely.
 *
 * Each one is a meal that was served and never recorded. They are kept until
 * someone acknowledges them, so the loss is visible rather than silent.
 */
export function expiredScans(): QueuedScan[] {
  const now = Date.now();
  return read().filter((entry) => now - entry.queuedAt >= MAX_QUEUE_AGE_MS);
}

/** Forgets stale entries, once staff have seen them. */
export function clearExpiredScans(): void {
  write(queuedScans());
}

export function enqueueScan(body: Record<string, unknown>): QueuedScan {
  const entry: QueuedScan = {
    id: `${Date.now()}-${Math.random().toString(36).slice(2, 10)}`,
    queuedAt: Date.now(),
    body,
  };
  write([...read(), entry]);
  return entry;
}

export function dequeueScan(id: string): void {
  write(read().filter((entry) => entry.id !== id));
}

export interface FlushResult {
  readonly synced: number;
  readonly failed: number;
}

/**
 * Replays everything buffered.
 *
 * A queued scan is removed when the server *answers* — including when it
 * answers `ALREADY_SERVED`, because that means the attendance row exists and
 * there is nothing left to sync. Only a transport failure keeps an entry
 * queued, so a legitimate denial cannot loop forever.
 */
export async function flushScanQueue(): Promise<FlushResult> {
  let synced = 0;
  let failed = 0;

  for (const entry of queuedScans()) {
    try {
      const response = await fetch("/api/qr/verify", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(entry.body),
      });

      if (response.ok) {
        dequeueScan(entry.id);
        synced++;
      } else if (response.status >= 400 && response.status < 500) {
        // The server understood and refused — a 401/403/400 will refuse
        // identically forever. Drop it rather than retry to exhaustion.
        dequeueScan(entry.id);
        failed++;
      } else {
        failed++;
      }
    } catch {
      // Still offline. Keep it for the next attempt.
      failed++;
    }
  }

  return { synced, failed };
}

File: mobile/lib/src/features/staff/scan_queue.dart
/// Scans buffered while the counter is offline (§6.4).
///
/// Counter Wi-Fi drops. When it does, the scan is buffered and replayed rather
/// than lost — the student has already walked away with their food, so the
/// attendance row must eventually exist.
///
/// Replay is safe because the server is idempotent: the
/// `UNIQUE (tenant_id, student_id, service_date, meal_slot)` constraint means a
/// replayed scan produces `ALREADY_SERVED` and zero additional rows. That is a
/// **success** from this queue's point of view — the record exists, which is all
/// it was trying to achieve.
///
/// Only a transport failure is ever queued. An HTTP denial is a decision the
/// server has already made, and queueing it would retry a legitimate refusal
/// forever — which is exactly why `/api/qr/verify` answers denials with 200.
library;

import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

/// Beyond this the meal is over and replaying would file the wrong service
/// date, so a stale entry must not be synced.
///
/// It must not be **discarded** either. These are students who already walked
/// away with their food; a tablet left offline overnight would otherwise erase a
/// whole service with nobody ever knowing. Stale entries are held back from sync
/// and surfaced to staff instead — see [expired].
const Duration kMaxQueueAge = Duration(hours: 6);

const int _maxEntries = 500;
const String _storageKey = 'campusmeals.scanQueue.v1';

class QueuedScan {
  const QueuedScan({
    required this.id,
    required this.queuedAt,
    required this.body,
  });

  /// Client-generated, so a retry of the same buffered scan is recognisable.
  final String id;
  final DateTime queuedAt;

  /// The exact verify request body, replayed unchanged.
  final Map<String, dynamic> body;

  Map<String, dynamic> toJson() => {
    'id': id,
    'queuedAt': queuedAt.millisecondsSinceEpoch,
    'body': body,
  };

  static QueuedScan? fromJson(Object? raw) {
    if (raw is! Map) return null;
    final id = raw['id'];
    final at = raw['queuedAt'];
    final body = raw['body'];
    if (id is! String || at is! int || body is! Map) return null;
    return QueuedScan(
      id: id,
      queuedAt: DateTime.fromMillisecondsSinceEpoch(at),
      body: body.cast<String, dynamic>(),
    );
  }
}

class ScanQueue {
  ScanQueue({SharedPreferences? prefs}) : _injected = prefs;

  final SharedPreferences? _injected;
  SharedPreferences? _prefs;

  Future<SharedPreferences> get _store async =>
      _prefs ??= _injected ?? await SharedPreferences.getInstance();

  Future<List<QueuedScan>> _read() async {
    try {
      final raw = (await _store).getString(_storageKey);
      if (raw == null) return [];
      final decoded = jsonDecode(raw);
      if (decoded is! List) return [];
      return decoded
          .map(QueuedScan.fromJson)
          .whereType<QueuedScan>()
          .toList(growable: false);
    } catch (_) {
      // Corrupt storage is treated as empty rather than throwing. A scanner
      // that will not start because of a malformed cache is worse than one that
      // has forgotten a queue it could not have read anyway.
      return [];
    }
  }

  Future<void> _write(List<QueuedScan> entries) async {
    try {
      // Keep the newest. If the cap is ever hit something is badly wrong, and
      // the most recent scans are the ones still inside their service window.
      final kept = entries.length > _maxEntries
          ? entries.sublist(entries.length - _maxEntries)
          : entries;
      await (await _store).setString(
        _storageKey,
        jsonEncode(kept.map((e) => e.toJson()).toList()),
      );
    } catch (_) {
      // Swallowed: a quota error must not take down the counter mid-service.
    }
  }

  Future<void> enqueue(Map<String, dynamic> body) async {
    final entries = [...await _read()];
    entries.add(
      QueuedScan(
        id: '${DateTime.now().millisecondsSinceEpoch}-${entries.length}',
        queuedAt: DateTime.now(),
        body: body,
      ),
    );
    await _write(entries);
  }

  /// Entries still inside their service window, and therefore safe to replay.
  Future<List<QueuedScan>> pending({DateTime? now}) async {
    final at = now ?? DateTime.now();
    return (await _read())
        .where((e) => at.difference(e.queuedAt) < kMaxQueueAge)
        .toList(growable: false);
  }

  /// Entries too old to replay — held, never deleted, and surfaced to staff.
  ///
  /// Each one is a student who ate. Silently dropping them would erase a
  /// service; replaying them would file it against the wrong day.
  Future<List<QueuedScan>> expired({DateTime? now}) async {
    final at = now ?? DateTime.now();
    return (await _read())
        .where((e) => at.difference(e.queuedAt) >= kMaxQueueAge)
        .toList(growable: false);
  }

  Future<void> remove(String id) async {
    final entries = (await _read()).where((e) => e.id != id).toList();
    await _write(entries);
  }

  /// Staff have been told about the stale entries and dismissed them.
  Future<void> clearExpired({DateTime? now}) async {
    final at = now ?? DateTime.now();
    await _write(
      (await _read())
          .where((e) => at.difference(e.queuedAt) < kMaxQueueAge)
          .toList(),
    );
  }
}

/// What to do with a queued scan after the server answered.
enum ReplayVerdict {
  /// The record exists. Dequeue.
  ///
  /// Includes `ALREADY_SERVED`: a replayed scan hitting the idempotency
  /// constraint is the queue succeeding, not failing.
  synced,

  /// The server understood and refused permanently. Dequeue — retrying a 4xx
  /// forever helps nobody.
  rejected,

  /// Server fault or no connection. Keep it and try later.
  keep,
}

/// Decides a queued scan's fate from the transport outcome.
///
/// Pure so the rule can be tested without a network: the difference between
/// dropping a student's meal record and retrying forever is one branch here.
ReplayVerdict verdictFor({required int? status}) {
  if (status == null) return ReplayVerdict.keep; // never reached the server
  if (status >= 200 && status < 300) return ReplayVerdict.synced;
  if (status >= 400 && status < 500) return ReplayVerdict.rejected;
  return ReplayVerdict.keep; // 5xx — the server's problem, not the scan's
}
